ext/ffi/lib.rs

```diff
@@ -535,322 +564,792 @@ where
+    },
+    NativeType::F32 => NativeValue {
+      f32_value: unsafe { cif.call::<f32>(fun_ptr, &call_args) },
+    },
+    NativeType::F64 => NativeValue {
+      f64_value: unsafe { cif.call::<f64>(fun_ptr, &call_args) },
+    },
+    NativeType::Pointer | NativeType::Function {} => NativeValue {
+      pointer: unsafe { cif.call::<*const u8>(fun_ptr, &call_args) },
+    },
+  })
+}
 
-  for (&native_type, value) in symbol
-    .parameter_types
-    .iter()
-    .zip(args.parameters.into_iter())
-  {
-    match native_type {
-      NativeType::Pointer => match value.as_u64() {
-        Some(idx) => {
-          let buf = buffers
-            .get(idx as usize)
-            .ok_or_else(|| {
-              generic_error(format!("No buffer present at index {}", idx))
-            })?
-            .unwrap();
-          native_values.push(NativeValue::buffer(buf.as_ptr()));
-        }
-        _ => {
-          let value = NativeValue::new(native_type, value)?;
-          native_values.push(value);
-        }
-      },
+struct RegisteredCallbackResource {
+  closure: libffi::middle::Closure<'static>,
+  info: *const CallbackInfo,
+}
+
+impl Resource for RegisteredCallbackResource {
+  fn name(&self) -> Cow<str> {
+    "registeredcallback".into()
+  }
+
+  fn close(self: Rc<Self>) {
+    let info = unsafe { Box::from_raw(self.info as *mut CallbackInfo) };
+    let isolate = unsafe { info.isolate.as_mut().unwrap() };
+    unsafe { v8::Global::from_raw(isolate, info.callback) };
+    unsafe { v8::Global::from_raw(isolate, info.context) };
+    drop(self)
+  }
+}
+
+struct CallbackInfo {
+  pub callback: NonNull<v8::Function>,
+  pub context: NonNull<v8::Context>,
+  pub isolate: *mut v8::Isolate,
+}
+
+unsafe extern "C" fn deno_ffi_callback(
+  cif: &libffi::low::ffi_cif,
+  result: &mut c_void,
+  args: *const *const c_void,
+  info: &CallbackInfo,
+) {
+  let isolate = &mut *info.isolate;
+  let callback = v8::Global::from_raw(isolate, info.callback);
+  let context = std::mem::transmute::<
+    NonNull<v8::Context>,
+    v8::Local<v8::Context>,
+  >(info.context);
+  IS_EVENT_LOOP_THREAD.with(|is_event_loop_thread| {
+    if !(*is_event_loop_thread.borrow()) {
+      // Call from another thread, PANIC IN THE DISCO!
+      todo!("Call from another thread");
+    }
+  });
+  // Call from main thread. If this callback is being triggered due to a
+  // function call coming from Deno itself, then this callback will build
+  // ontop of that stack.
+  // If this callback is being triggered outside of Deno (for example from a
+  // signal handler) then this will either create an empty new stack if
+  // Deno currently has nothing running and is waiting for promises to resolve,
+  // or will (very incorrectly) build ontop of whatever stack exists.
+  // The callback will even be called through from a `while (true)` liveloop, but
+  // it somehow cannot change the values that the loop sees, even if they both
+  // refer the same `let bool_value`.
```

Comment:
I do not follow. Can you elaborate?

Reply:
Ah, this has to do with the sick example I showed in Discord that used a signal handler on the FFI Rust side to call the registered callback.

So a code like this:
```js
let shouldContinueLoop = true;
const callback = () => {
  console.log("Stop looping!");
  shouldContinueLoop = false;
};
const regCb = new Deno.RegisteredCallback({ parameters: [], result: "void" }, callback);
dylib.symbols.store_callback(regCb);
dylib.symbols.call_stored_callback_after_delay(); // Sets up a SIGALRM interrupt handler that calls the stored callback, and then sets up a SIGALRM interrupt signal to be fired in 2 seconds time.

let count = 0;
while(shouldContinueLoop) {
  if (count++ == 10 ** 6) {
  console.log("No I won't!");
 }
}
```
will output:
> Stop looping!
> No I won't!
and will never exit.

If you add a `performance.now()` call into the loop, then it will exit properly.
If you add a `Math.random()` call into the loop, then it will panic.
Also, if you add a `console.trace()` to the callback you'll see the callback emanating from inside the while-loop.
So, if an interrupt handler is setup to call back into JS and Deno currently has an active stack, the callback will simply step on top of that stack and do its work. That does not mean that the data that Deno (or V8, rather) uses will be affected as expected.

This is, I think, something that should be carefully documented but trying to fight against this would probably not be a good idea, as I don't think that at present we have the tools to do anything except accept this.